src/main/java/com/gregtechceu/gtceu/common/item/ColorSprayBehaviour.java

```diff
@@ -329,28 +342,40 @@ private boolean tryPaintSpecialBlock(Player player, Level world, BlockPos pos, B
 
         // MTE special case
         BlockEntity be = world.getBlockEntity(pos);
-        if (be instanceof IMachineBlockEntity machineBe) {
+        if (be instanceof IMachineBlockEntity machineBe && first instanceof IMachineBlockEntity og) {
             MetaMachine mte = machineBe.getMetaMachine();
-            if (mte != null) {
-                if (mte.getPaintingColor() != this.color.getTextColor()) {
+            MetaMachine ogMte = og.getMetaMachine();
+            if (mte != null && ogMte != null) {
+                if (mte.getPaintingColor() != this.color.getTextColor() &&
+                        mte.getPaintingColor() == ogMte.getPaintingColor()) {
                     mte.setPaintingColor(this.color.getTextColor());
                     return true;
                 } else return false;
             }
         }
 
         // PipeBlockEntity special case
-        if (be instanceof PipeBlockEntity<?, ?> pipe) {
-            if (pipe.getPaintingColor() != this.color.getTextColor()) {
+        if (be instanceof PipeBlockEntity<?, ?> pipe && first instanceof PipeBlockEntity<?, ?> og) {
+            var side = getPaintDirection(player).getOpposite();
+            if (!first.getBlockPos().equals(be.getBlockPos()) && (!pipe.isConnected(side))) {
+                return false;
+            }
+            if (pipe.getPaintingColor() != this.color.getTextColor() &&
+                    pipe.getPaintingColor() == og.getPaintingColor()) {
                 pipe.setPaintingColor(this.color.getTextColor());
                 return true;
             } else return false;
         }
 
         if (GTCEu.isAE2Loaded()) {
-            if (be instanceof CableBusBlockEntity cable) {
+            if (be instanceof CableBusBlockEntity cable && first instanceof CableBusBlockEntity og) {
+                var side = getPaintDirection(player).getOpposite();
+                if (!first.getBlockPos().equals(be.getBlockPos()) &&
+                        (cable.getPart(side) != null || cable.getCableConnectionType(side) == AECableType.NONE)) {
```

Comment:
If I am reading this right, for each block, it checks if near side has a cable part, and if so, stops before painting that block. I *think* you also want to check is the *previous* block has a cable part on the far side.

Reply:
for both ae2 cables and gt pipes it's almost the same thing iirc, if the previous cable has a part on the far side the near side of the current one will either be disconnected or have another part.

Reply:
With no knowledge of how AE represents things internally, is that true of ME P2P tunnel? Certainly, from a logical perspective, the cable opposite the tunnel is connected to the tunnel. I don't know what that means in terms of `getCableConnectionType` though.

Reply:
you're right, p2p is an edge case, I'll see how it can be handled

Reply:
the right way to do it would probably be a refactor of the behaviour to use some custom collector logic for each block paintable block type, that would also help if we want to paint the connected network instead of just a straight line, I'll give it a shot later

Reply:
fixed